src/libraries/Microsoft.PowerFx.Core/Public/CheckResult.cs

```diff
@@ -17,115 +19,413 @@
 namespace Microsoft.PowerFx
 {
     /// <summary>
-    /// Result of binding an expression. 
+    /// Holds work such as parsing, binding, error checking done on a single expression. 
+    /// Different options require different work. 
+    /// Tracks which work is done so that it is not double repeated.
     /// </summary>
     public class CheckResult : IOperationStatus
     {
-        /// <summary> 
-        /// Return type of the expression. Null if type can't be determined. 
+        #region User Inputs
+        // Captured from user, set once via a Set*() function. 
+
+        /// <summary>
+        /// The source engine this was created against. 
+        /// This is critical for calling back to populate the rest of the results. 
         /// </summary>
-        public FormulaType ReturnType { get; set; }
+        private readonly Engine _engine;
+       
+        // The raw expression test. 
+        private string _expression;
+        
+        private ParserOptions _parserOptions;
+
+        // Information for binding. 
+        private bool _setBindingCalled;
+        internal ReadOnlySymbolTable _symbols; // can be null
+        private VersionHash _symbolHash; // hash of _symbols at time of assignment
+
+        #endregion
+
+        [Obsolete("Use public constructor")]
+        internal CheckResult()
+        {
+        }
 
         /// <summary>
-        /// Names of fields that this formula uses. 
-        /// null if unavailable.  
-        /// This is only valid when <see cref="IsSuccess"/> is true.
+        /// Initializes a new instance of the <see cref="CheckResult"/> class.
         /// </summary>
-        public HashSet<string> TopLevelIdentifiers { get; set; }
+        /// <param name="source">Engine used to handle Apply operations.</param>
+        public CheckResult(Engine source)

```

Comment:
>CheckResult [](http://example.com/codeflow?start=15&length=11)

Should we initialize _errors here?

Reply:
it's initialized by the field initializer. (that runs as part of the ctor)